address.md

```diff
@@ -0,0 +1,300 @@
+
+```
+|------------|----------|---------------------|----------|
+|  network   | protocol |      payload        | checksum |
+|------------|----------|---------------------|----------|
+| 'f' or 't' |     2    |  blake2b-160(TODO)  |  4 bytes |
+                  base32[................................]
+```
+
+###### Protocol 3: BLS
+
+**Protocol 3** addresses represent BLS public encryption keys. The payload field contains the BLS public key.
+
+**Bytes**
+
+```
+|----------|---------------------|
+| protocol |        payload      |
+|----------|---------------------|
+|    1     | 48 byte BLS PubKey  |
+```
+
+**String**
+
+```
+|------------|----------|---------------------|----------|
+|  network   | protocol |      payload        | checksum |
+|------------|----------|---------------------|----------|
+| 'f' or 't' |     3    |  48 byte BLS PubKey |  4 bytes |
+                  base32[................................]
+```
+
+#### Payload
+
+The payload represents the data specified by the protocol. All payloads except the payload of the ID protocol are [base32](https://tools.ietf.org/html/rfc4648) encoded using the lowercase alphabet when seralized to their human readable format.
+
+#### Checksum
+
+The checksum is calculated using blake2b-4 hash of the addresses protocol, and payload bytes. Checksums are base32 encoded  and only used when encoding and decoding an address to its human readable format -- a string. Checksums will never appear on chain.
+
+
+#### Expected Methods
+
+All implementations in Filecoin must have methods for creating, encoding, and decoding addresses in addition to checksum creation and validation. The follwing is a golang version of the Address Interface:
+
+```go
+type Address interface {
+    New(p byte, payload []byte) Address
+    Encode(network Network, a Adress) string
+    Decode(s string) Address
+    Checksum(a Address) []byte
+    ValidateChecksum(a Address) bool
+}
+```
+##### New()
+
+New returns an Address for the specified protocol encapsulating corresponding payload. New fails for unknown protocol.
+
+```go
+func New(protocol byte, payload []byte) Address {
+	if protocol < SECP256K1 || protocol > BLS {
+		Fatal(ErrUnknownType)
+	}
+    return Address{
+        Protocol: protocol,
+        Payload:  payload,
+    }
+}
+```
+
+##### Encode()
+
+Software encoding a Filecoin address must:
+
+- produce an address encoded to a known network
+- produce an address encoded to a known protocol
+- produce an address with a valid checksum
+
+Encodes an Address as a string, prepending the network prefix, calculating the checksum and encoding the payload and checksum to [base32](https://tools.ietf.org/html/rfc4648).
+
+```go
+func Encode(network string, a Address) string {
+	if network != "f" && network != "t" {
+        Fatal("Invalid Network")
+	}
+	
+    switch a.Protocol {
+    	case SECP256K1:
```

Comment:
note that if using go-like syntax, this is broken. This function will no-op for Secp and Actor. either explicitly say 'fallthrough' or do `case BLS, Actor, SECP256k1:`